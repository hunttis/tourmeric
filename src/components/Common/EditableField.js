import React, { Component } from 'react';
import firebase from 'firebase/app';
import { Translate } from 'react-localize-redux';
import _ from 'lodash';
import PropTypes from 'prop-types';

export default class EditableField extends Component {

  delayedSave = _.debounce((path, value) => {
    firebase.update(path, value);
    this.setState({ saved: true, editing: false });
    this.delayedNormalize();
  }, 300)

  delayedNormalize = _.debounce(() => {
    this.setState({ saved: false, editing: false });
  }, 2000);

  state = { saved: false, editing: false }

  handleChange(path, targetName, value) {
    this.setState({ editing: true, saved: false });
    this.delayedSave(path, { [targetName]: value });
  }

  render() {
    const {
      labelContent, placeHolder, defaultValue, path, targetName, inputType = 'text', inputClasses, leftIcon,
    } = this.props;
    const { saved, editing } = this.state;

    return (
      <div className="editablefield field is-horizontal">
        {labelContent &&
        <div className="field-label is-normal">
          <label className="label">
            <Translate id={labelContent} />
          </label>
        </div>
        }
        <div className="field-body">
          <div className="field">
            <p className={`control is-expanded ${leftIcon && 'has-icons-left'} has-icons-right`}>

              <Translate>
                {translate => (<input
                  type={inputType}
                  className={`input ${saved && 'is-success'} ${editing && 'is-warning'} ${(!editing && !saved) && 'is-normal'} ${inputClasses}`}
                  placeholder={translate(placeHolder)}
                  defaultValue={defaultValue}
                  onChange={event => this.handleChange(path, targetName, event.target.value)}
                />)
                }
              </Translate>
              {leftIcon && <span className="icon is-small is-left"><i className={`fas fa-${leftIcon}`} /></span>}
              {saved && <span className="icon is-small is-right has-text-success"><i className="fas fa-check-circle" /></span>}
              {editing && <span className="icon is-small is-right has-text-warning"><i className="fas fa-pencil-alt" /></span>}
            </p>
          </div>
        </div>
      </div>
    );
  }
}

EditableField.propTypes = {
  labelContent: PropTypes.string,
  placeHolder: PropTypes.string,
  defaultValue: PropTypes.string,
  path: PropTypes.string,
  targetName: PropTypes.string,
  inputType: PropTypes.string,
  inputClasses: PropTypes.string,
  leftIcon: PropTypes.string,
};
