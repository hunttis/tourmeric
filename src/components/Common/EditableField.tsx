import React, { Component } from 'react';
import firebase from 'firebase/app';
import { Translate } from 'react-localize-redux';
import _ from 'lodash';

interface Props {
  labelContent: string;
  placeHolder: string;
  defaultValue: string;
  path: string;
  targetName: string;
  inputType?: string;
  inputClasses?: string;
  leftIcon?: string;
}

interface State {
  editing: boolean;
  saved: boolean;
}

export default class EditableField extends Component<Props, State> {

  delayedSave = _.debounce((path, value) => {
    firebase.update(path, value);
    this.setState({ saved: true, editing: false });
    this.delayedNormalize();
  }, 300)

  delayedNormalize = _.debounce(() => {
    if (!this.unmounting) {
      this.setState({ saved: false, editing: false });
    }
  }, 2000);

  state = { saved: false, editing: false }

  unmounting: boolean = false;

  componentWillUnmount() {
    this.unmounting = true;
  }

  handleChange(path: string, targetName: string, value: string) {
    this.setState({ editing: true, saved: false });
    this.delayedSave(path, { [targetName]: value });
  }

  render() {
    const {
      labelContent, placeHolder, defaultValue, path, targetName, inputType = 'text', inputClasses = '', leftIcon,
    } = this.props;
    const { saved, editing } = this.state;

    return (
      <div className="editablefield field is-horizontal">
        {labelContent &&
          <div className="field-label is-normal">
            <label className="label">
              <Translate id={labelContent} />
            </label>
          </div>
        }
        <div className="field-body">
          <div className="field">
            <p className={`control is-expanded ${leftIcon && 'has-icons-left'} has-icons-right`}>

              <Translate>
                {(translate: any) => (<input
                  type={inputType}
                  className={`input ${saved && 'is-success'} ${editing && 'is-warning'} ${(!editing && !saved) && 'is-normal'} ${inputClasses}`}
                  placeholder={translate(placeHolder)}
                  defaultValue={defaultValue}
                  onChange={event => this.handleChange(path, targetName, event.target.value)}
                />)
                }
              </Translate>
              {leftIcon && <span className="icon is-small is-left"><i className={`fas fa-${leftIcon}`} /></span>}
              {saved && <span className="icon is-small is-right has-text-success"><i className="fas fa-check-circle" /></span>}
              {editing && <span className="icon is-small is-right has-text-warning"><i className="fas fa-pencil-alt" /></span>}
            </p>
          </div>
        </div>
      </div>
    );
  }
}
