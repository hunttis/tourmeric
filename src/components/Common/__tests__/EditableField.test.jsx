import React from 'react';
import { shallow } from 'enzyme';
import EditableField from '../EditableField';

describe('EditableField Tests', () => {

  it('Compares EditableField to snapshot', () => {

    const editableField = shallow(<EditableField
      labelContent="label"
      placeHolder="placeholder"
      defaultValue="defaultvalue"
      path="path"
      targetName="target"
      inputType="input"
      inputClasses="inputClasses"
      leftIcon="lefticon"
    />);

    expect(editableField.exists()).toBe(true);
    expect(editableField).toMatchSnapshot();
  });

  // TODO: Write a test for the save and normalize functions

});
