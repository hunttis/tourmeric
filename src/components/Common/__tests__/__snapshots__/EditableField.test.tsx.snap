// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableField Tests Compares EditableField to snapshot 1`] = `
<div
  className="editablefield field is-horizontal"
>
  <div
    className="field-label is-normal"
  >
    <label
      className="label"
    >
      <Translate
        id="label"
      />
    </label>
  </div>
  <div
    className="field-body"
  >
    <div
      className="field"
    >
      <p
        className="control is-expanded has-icons-left has-icons-right"
      >
        
          <Component />
        
        <span
          className="icon is-small is-left"
        >
          <i
            className="fas fa-lefticon"
          />
        </span>
      </p>
    </div>
  </div>
</div>
`;
